keps/sig-scheduling/5055-dra-device-taints-and-tolerations/README.md

```diff
@@ -399,34 +429,109 @@ As with Taint, these structs get duplicated to document DRA specific
+corresponds to
+[v1.Toleration.ToleratesTaint](https://github.com/kubernetes/kubernetes/blob/85734ac6b38e29a9a390520e7a5b6de1fbf5ff6b/staging/src/k8s.io/api/core/v1/toleration.go#L29-L57).
+
+To enable setting of taints without reconfiguring DRA drivers, the
+cluster-scoped DeviceTaintRule gets added. The scheduler must merge these
+additional taints with the ones provided by
+the DRA drivers on-the-fly while it gathers information about available
+devices. This could be done each time a scheduling cycle starts, but it
+would be repetitive work. Instead, a
+[ResourceSlice tracker](https://github.com/kubernetes/kubernetes/blob/85734ac6b38e29a9a390520e7a5b6de1fbf5ff6b/staging/src/k8s.io/dynamic-resource-allocation/resourceslice/tracker/tracker.go#L56-L59)
+reacts to informer events for ResourceSlice and DeviceTaintRule and
+maintains a set of updated ResourceSlices which also contain the taints
+set via a DeviceTaintRule. The tracker provides the API of an informer
+and thus can be used as a drop-in replacement for a ResourceSlice
+informer.
 
 ```Go
-type DevicePatch struct {
-    ...
+// DeviceTaintRule adds one taint to all devices which match the selector.
+// This has the same effect as if the taint was specified directly
+// in the ResourceSlice by the DRA driver.
+type DeviceTaintRule struct {
+    metav1.TypeMeta
+    // Standard object metadata
+    // +optional
+    metav1.ObjectMeta
 
-    // If specified, the device's taints. Taints with unique key and effect
-    // get added to the set of taints of the device. When key and effect
-    // are used in multiple places, the same precedence rules as for attributes apply
-    // (see the priority field).
+    // Spec specifies the selector and one taint.
     //
-    // The maximum number of tolerations is 16.
+    // Changing the spec automatically increments the metadata.generation number.
+    Spec DeviceTaintRuleSpec
+
+    // ^^^
+    // A spec gets added because adding a status seems likely.
+    // Such a status could provide feedback on applying the
+    // eviction and/or statistics (number of matching devices,
+    // affected allocated claims, pods remaining to be evicted,
+    // etc.).
+}
+
+// DeviceTaintRuleSpec specifies the selector and one taint.
+type DeviceTaintRuleSpec struct {
+    // DeviceSelector defines which device(s) the taint is applied to.
+    // All selector criteria must be satified for a device to
+    // match. The empty selector matches all devices. Without
+    // a selector, no devices are matches.
     //
-    // This is an alpha field and requires enabling the DRADeviceTaints
-    // feature gate.
+    // +optional
+    DeviceSelector *DeviceTaintSelector
+
+    // The taint that gets applied to matching devices.
+    //
+    // +required
+    Taint DeviceTaint
```

Comment:
Any reason why it's not slice?

Reply:
Yes: if you need more than one, create multiple DeviceTaintRules. The API is simpler this way.